broadinstitute/single_cell_portal_core
# Updating base Docker image, Rails to 6.1.7.1, automating local dev tasks (SCP-4938)

#### BACKGROUND & CHANGES
This updates to the latest release of the base Docker image `gcr.io/broad-singlecellportal-staging/rails-baseimage:2.2.0`.  This contains normal security patches to the underlying Ubuntu 20.04 OS, as well as upgrading the Ruby version to 3.1.3.  Also, this updates the Rails version to `6.1.7.1` to address CVE-2023-22794.

Additionally, it was discovered while working on this update that the current `docker-compose` automation does not work terribly well when changes have been made to the upstream Docker image or locally to the Gemfile.  It necessitated rebuilding the image locally and tagging it correctly with `gcr.io/broad-singlecellportal-staging/single-cell-portal:development` before calling `bin/docker-compose-setup.sh`.  This borrows from some recent work in [scp-ingest-pipeline](https://github.com/broadinstitute/scp-ingest-pipeline/pull/290) to automatically pull in updates to the Docker image, as well as allowing the user to specify alternate image tags, or completely rebuild the image locally and use that instead if changes have been detected.  The latter option allows testing of this PR without any additional steps.  

Lastly, this also automates rebuilding the development image locally in CI for the same reasons as any upstream changes will cause builds to fail since the image is out of sync.  This will only happen if a change is detected to the base `Dockerfile` to keep build times as short as possible.

#### MANUAL TESTING
1. Pull this branch
2. From project source, run `bin/docker-compose-setup.sh` (this should automatically detect that the Dockerfile has changed and rebuild the image locally)
3. After the image builds and the Rails server & vite start, confirm you can bring up the home page
4. (Optional) Attach to the Rails console and confirm that you see the version of Rails has updated to `6.1.7.1` by running `docker exec -it single_cell bin/rails c` in a separate terminal:
```
Loading development environment (Rails 6.1.7.1)
irb(main):001:0> 
```
